crates/isograph_compiler/src/watch.rs

```diff
@@ -1,47 +1,55 @@
 use colored::Colorize;
 use isograph_config::CompilerConfig;
-use notify::{Error, RecommendedWatcher, RecursiveMode, Watcher};
+use notify::{
+    event::{CreateKind, ModifyKind, RemoveKind, RenameMode},
+    Error, EventKind, RecommendedWatcher, RecursiveMode,
+};
 use notify_debouncer_full::{
-    new_debouncer, DebounceEventResult, DebouncedEvent, Debouncer, FileIdMap,
+    new_debouncer, DebounceEventResult, DebouncedEvent, Debouncer, RecommendedCache,
 };
 use std::{path::PathBuf, time::Duration};
 use tokio::{runtime::Handle, sync::mpsc::Receiver, task::JoinError};
 
-use crate::batch_compile::compile_and_print;
+use crate::{
+    batch_compile::print_result, compiler_state::CompilerState, with_duration::WithDuration,
+};
+
+const MAX_CHANGED_FILES: usize = 100;
 
 pub async fn handle_watch_command(
-    config: CompilerConfig,
+    config_location: PathBuf,
 ) -> Result<Result<(), Vec<Error>>, JoinError> {
-    let _ = compile_and_print(&config);
-
+    let mut state = CompilerState::new(config_location);
     let (mut rx, mut watcher) = create_debounced_file_watcher();
+    configure_watcher(&mut watcher, &state.config);
 
-    // We need to watch a few things: the schema, extensions, and project root
-    watcher
-        .watcher()
-        .watch(&config.project_root, RecursiveMode::Recursive)
-        .expect("Failure when watching project root");
-    watcher
-        .watcher()
-        .watch(&config.schema, RecursiveMode::Recursive)
-        .expect("Failing when watching schema");
-    for extension in &config.schema_extensions {
-        watcher
-            .watcher()
-            .watch(extension, RecursiveMode::Recursive)
-            .expect("Failing when watching schema extension");
-    }
+    eprintln!("{}", "Starting to compile.".cyan());
+    let _ = print_result(WithDuration::new(|| state.compile()));
 
     tokio::spawn(async move {
         while let Some(res) = rx.recv().await {
             match res {
                 Ok(events) => {
-                    if any_modified_path_is_outside_artifact_directory(
-                        &events,
-                        &config.artifact_directory,
-                    ) {
-                        eprintln!("{}", "File changes detected.".cyan());
-                        let _ = compile_and_print(&config);
+                    if let Some(changes) = categorize_and_filter_events(&events, &state.config) {
+                        let result = if has_config_changes(&changes) {
+                            eprintln!(
+                                "{}",
+                                "Config change detected. Starting a full compilation.".cyan()
+                            );
+                            state.reload_config();
+                            reconfigure_watcher(&mut watcher, &state.config);
```

Comment:
it might be cleaner to not have a `[re]configure_watcher` functions, and instead just do

```
let (new_rx, new_watcher) = create_debounced_file_watcher(&config);
*rx = new_rx;
*watcher = new_watcher;
*state = create_entirely_new_state(); // just reread everything
```

what do you think? IMO this is easier to reason about, instead of methods that take `&mut self`

Reply:
Agreed. I tried to do this first, but had some issues. But yeah, now I see how to implement it, thanks.